### ISSUE TITLE
MathEx.squaredDistance is slow

### ISSUE THREAD

**Issue by lukehutch** [haifengl/smile#597]
**Describe the bug**
When running a very large number of `KDTree` queries, the code spends 80% of its time in `MathEx.squaredDistance`.

**Expected behavior**
Squared distance should not take that long.

**Additional context**
JDK 15, Linux

I suggest adding the following optimization for the common 2, 3 and 4 dimensional cases (the kd-tree algorithm is not very efficient for >5 dimensions anyway, as it quickly degrades to linear search):

```java
    public static double squaredDistance(float[] x, float[] y) {
        if (x.length != y.length) {
            throw new IllegalArgumentException("Input vector sizes are different.");
        }

        double sum = 0.0;
        for (int i = 0; i < x.length; i++) {
            sum += sqr(x[i] - y[i]);
        }

        return sum;
    }
```

should be 

```java
    public static double squaredDistance(float[] x, float[] y) {
        if (x.length != y.length) {
            throw new IllegalArgumentException("Input vector sizes are different.");
        } else if (x.length == 2) {
            double d0 = (double) x[0] - (double) y[0];
            double d1 = (double) x[1] - (double) y[1];
            return d0 * d0 + d1 * d1;
        } else if (x.length == 3) {
            double d0 = (double) x[0] - (double) y[0];
            double d1 = (double) x[1] - (double) y[1];
            double d2 = (double) x[2] - (double) y[2];
            return d0 * d0 + d1 * d1 + d2 * d2;
        } else if (x.length == 4) {
            double d0 = (double) x[0] - (double) y[0];
            double d1 = (double) x[1] - (double) y[1];
            double d2 = (double) x[2] - (double) y[2];
            double d3 = (double) x[3] - (double) y[3];
            return d0 * d0 + d1 * d1 + d2 * d2 + d3 * d3;
        }

        double sum = 0.0;
        for (int i = 0; i < x.length; i++) {
            sum += sqr(x[i] - y[i]);
        }

        return sum;
    }
```

Or maybe you could use a `switch` statement (these can be executed fast using a jump table), and then you could provide a fast path for up to maybe 10 dimensions or something.

The function `double squaredDistance(double[] x, double[] y)` also needs to be modified in the same way.

Note that there is also a numerical loss of precision bug in the current code:

```java
            sum += sqr(x[i] - y[i]);
```

should be 

```java
            sum += sqr((double) x[i] - (double) y[i]);
```

otherwise subtracting a very small number from a very large number (or vice versa) will lose precision. In fact you lose a digit of precision for every order of magnitude in difference in the two numbers. This bug affects several methods.

PS I also came here to report a major inefficiency for `KDTree.buildNode`, where `upperBound` and `lowerBound` were being allocated for every recursion frame -- but it looks like you fixed that recently. Also I was going to report a bug in `KDTree.buildNode` that could lead to an infinite loop, but it looks like you fixed that recently too by making the node a leaf node `if (size == 0 || size == node.count)`. Thanks for these recent fixes!

**Comment by haifengl** [comment]
Thanks for the suggestion. I tried it but the benchmark of a 3-dimensional data doesn't show notable difference. I believe JVM already do loop unrolling optimization automatically.

**Comment by lukehutch** [comment]
OK. I'll let you know if I can produce a benchmark that shows a noticeable difference. However, see also the comment about loss of precision.

**Comment by lukehutch** [comment]
I wrote some quick benchmark code that shows that the optimized version is 13% faster than the unoptimized version for 3 dimensions (which is significant for my usecase). The performance gap increases from a few percent to 13% (or potentially more) when the query points are relatively far from the points in the kd-tree (since that requires more backtracking during kd-tree search). I suspect the exact speedup will be dependent upon the processor used, and the distribution of points in the kd-tree and query cloud.

**Comment by haifengl** [comment]
How large is your data? How many queries were in the benchmark? What's the time?

**Comment by lukehutch** [comment]
Here's the code. With `offset = size` (which causes the cube of kd-tree points to be offset from the cube of query points by the width of the cube, causing longer lookup times), lookup time for the optimized version was about 14.2 seconds on a Ryzen 3950X on one core, and for the original `KDTree` it was 13% higher than that. Construction time was similar for both. The commented-out code is the calls to the 3D-optimized version of the `KDTree` and `Neighbor` classes. I didn't play with the parameters or the distribution of points much at all, but I bet you could get a higher speedup than 13%, e.g. maybe by setting all the kd-tree points on the surface of a sphere, and then querying random points within a cube twice the size of the sphere (since there would be a lot of cases where the distance to the closest point would be much larger than the distance between the k nearest neighbors, which causes search to degrade to linear).

```java
    public static void main(String[] args) {
        int N = 1_000_000;
        int size = 100;
        int numTests = 1_000;
        double offset = size;

        Random rand = new Random(1);

        double[][] coords = new double[N][3];
        Integer[] vals = new Integer[N];
        for (int i = 0; i < N; i++) {
            coords[i][0] = rand.nextDouble() * size - size / 2;
            coords[i][1] = rand.nextDouble() * size - size / 2;
            coords[i][2] = rand.nextDouble() * size - size / 2;
            vals[i] = i;
        }
        long startTime0 = System.nanoTime();
        KDTree<Integer> kdt = new KDTree<>(coords, vals);
        // DoubleKDTree3D<Integer> kdt = new DoubleKDTree3D<Integer>(coords, vals);
        System.out.println("Construction: " + (System.nanoTime() - startTime0) * 1.0e-9 + " sec");

        double[] q = new double[3];
        long startTime1 = System.nanoTime();
        for (int i = 0; i < numTests; i++) {
            q[0] = rand.nextDouble() * size - size / 2 + offset;
            q[1] = rand.nextDouble() * size - size / 2 + offset;
            q[2] = rand.nextDouble() * size - size / 2 + offset;
            Neighbor<double[], Integer>[] neighbors = kdt.knn(q, 10);
            Neighbor<double[], Integer> nearest = kdt.nearest(q);
            //            DoubleNeighbor3D<Integer>[] neighbors = kdt.knn(q, 10);
            //            DoubleNeighbor3D<Integer> nearest = kdt.nearest(q);
        }
        System.out.println("Lookups: " + (System.nanoTime() - startTime1) * 1.0e-9 + " sec");

        //        Construction: 0.8377116240000001 sec
        //        Lookups: 13.746584138000001 sec

        // Construction: 0.812098447 sec
        // Lookups: 14.222086729 sec

    }
```

**Comment by haifengl** [comment]
In your benchmark, none of query points are in the same range of KD-Tree data. This is very very rare, at least in machine learning use cases.  Without offset, the query time is only 0.5%. Therefore, I think that this benchmark doesn't really reflect the effect of loop unrolling.

**Comment by lukehutch** [comment]
Actually the whole point of that offset is to demonstrate that loop unrolling (and function inlining) is not happening, or not optimally. Without separating the query and target points by some distance, the algorithm spends a smaller percentage of the time in distance calculation compared to doing other work.

As far as whether or not this is common in machine learning -- it totally depends on the density of training data in the local neighborhood, and it also depends on whether you are interpolating or extrapolating. The key factor in whether a kd-tree search is closer to logarithmic or linear in behavior is the ratio of distance from query point to nearest neighbor vs. the average distance between the k nearest neighbors. In my application, it is quite common for behavior to degrade from logarithmic towards linear.

Again though, even if you don't want to fix this, please consider fixing the loss of numerical precision that I pointed out.

**Comment by haifengl** [comment]
precision loss protection is in the code. But I don't really think that it would be an issue for distance calculation in general since vectors are not long enough to see the impact. beside, KDTree use double[].

**Comment by lukehutch** [comment]
Oh, you're right, sorry. I forgot that I'm using a modified version of `KDTree` that uses `float` for point coordinates but `double` for distances. Maybe that's something you would be interested in adding: halving the precision actually doubled the speed of my KDT code.
